### ISSUE TITLE
Невереный сертификат

### ISSUE THREAD

**Issue by vadimkara** [krlvm/PowerTunnel#115]
![image](https://user-images.githubusercontent.com/7599876/182077977-391f556a-b71e-4cc7-8b62-80a97796acb6.png)

На некторых сайтх при переходе по ссылке на субдомены программа выдает сертификат не на тот днс что указан в ссылке

![image](https://user-images.githubusercontent.com/7599876/182078086-f56bc286-0cf3-4655-89f7-5a28781f3155.png)

**Comment by krlvm** [comment]
Эта проблема связана с Cloudflare, когда модификация SNI выключена, она отсутствует. Думаю, что можно с этим сделать.

**Comment by pundoo** [comment]
> This problem is related to Cloudflare, when the SNI modification is disabled, it is missing. I think what can be done about it.

Any workaround for this? I'm facing the same issue.

**Comment by pundoo** [comment]
> Эта проблема связана с Cloudflare, когда модификация SNI выключена, она отсутствует. Думаю, что можно с этим сделать.

Any workaround for this? I'm facing the same issue.

**Comment by pundoo** [comment]
> Эта проблема связана с Cloudflare, когда модификация SNI выключена, она отсутствует. Думаю, что можно с этим сделать.

I'm facing same issue. Any workaround for this (with sni, not chuck)?

**Comment by krlvm** [comment]
Not yet, but you can disable these warnings in Chrome (this is not a good way though)

**Comment by pundoo** [comment]
> Not yet, but you can disable these warnings in Chrome (this is not a good way though)

Disabling warnings don't seems to work. These sites are basically behind CDNs, that strictly requires the correct SNI field to serve the right certificate (and the right server to connect to?). Faking SNI simply breaks this mechanism.

I don't know (in-depth) exactly how the HTTPS connection establishes but, is it somehow possible to generate the (self-signed) cert from client request SNI/host (Fiddler does it), so that the CN matches and the handshake is successful. Then after that, somehow tell server the actual site we're trying to connect to via host header or something?

**Comment by krlvm** [comment]
> Disabling warnings don't seems to work. These sites are basically behind CDNs, that strictly requires the correct SNI field to serve the right certificate (and the right server to connect to?). Faking SNI simply breaks this mechanism.

Yes, that's the problem. IIRC I wrote a detailed explaination in other thread some time ago.

> I don't know (in-depth) exactly how the HTTPS connection establishes but, is it somehow possible to generate the (self-signed) cert from client request SNI/host (Fiddler does it), so that the CN matches and the handshake is successful. Then after that, somehow tell server the actual site we're trying to connect to via host header or something?

MITM is done by LittleProxy and its extension LittleProxy-mitm. Certificate generation is requested [here](https://github.com/krlvm/PowerTunnel/blob/c029df7d58920d3043dc05d52995170752adbd6a/littleproxy/src/main/java/org/littleshoot/proxy/impl/ProxyToServerConnection.java#L856), and it the certificates changes then, which is the problem. I tried to debug this previously, but I did not succeed. I will try again today.

**Comment by pundoo** [comment]
>MITM is done by LittleProxy and its extension LittleProxy-mitm. Certificate generation is requested [here](https://github.com/krlvm/PowerTunnel/blob/c029df7d58920d3043dc05d52995170752adbd6a/littleproxy/src/main/java/org/littleshoot/proxy/impl/ProxyToServerConnection.java#L856), and it the certificates changes then, which is the problem. I tried to debug this previously, but I did not succeed. I will try again today.

Did some read up and seems like it should be fairly simple. We just have to get the SNI/host from the client request and use it as the COMMON_NAME (CN) for self generated certificate (instead of using CN received from server). I don't know Java that much, otherwise I'd test this approach..

**Comment by krlvm** [comment]
Yes that's it. Though I don't see a way to obtain it yet.
